qualtran/bloqs/comparison_gates.ipynb

```diff
@@ -0,0 +1,957 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "less_than_circuit = cirq.Circuit(less_than(classical_number, quantum_number, z))\n",
+    "less_than_circuit"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Now we check individual possiblities.\n",
+    "check_each_possibility(less_than_circuit, quantum_number, [z])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Finally we check the uniform super position.\n",
+    "check_uniform_superposition(less_than_circuit, quantum_number, [z])"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**And as before notice that the ancilla qubit are always clean at the end of execution and that the input qubits are not affected**"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "qa-90pTK0Spu"
+   },
+   "source": [
+    "## Improving the constant\n",
+    "\n",
+    "$$\\renewcommand{\\ket}[1]{|#1\\rangle}$$\n",
+    "The implementation above has T complexity of exactly $4n$ since there are exactly $n$ And gates each uses $4$ Ts. Note however that the first of them is not actually needed since one of its inputs is in the $\\ket{1}$ state so it collapses to either identity or `cirq.X` depending on the most significant bit of $B$. This gives a T complexity of $4(n-1) = 4n - 4$."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "MDysRmBd1LFw"
+   },
+   "source": [
+    "## Drawback\n",
+    "This decomposition has the currently optimal T complexity for the quantum-classical case however it has linear depth. We will see later a way that has logarithmic depth but with $6n + \\mathcal{O}(1)$ T complexity."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## The Quantum-Quantum Comparator $8n + \\mathcal{O}(1)$ T gates\n",
+    "\n",
+    "For the quantum-classical case we scanned the numbers sequentially from left to right to compute the comparison result, a similar strategy can be use to create a decomposition that has $12n - 8$ T complexity and linear depth. We won't go in depth into that way but refer the reader to  my incomplete [PR](https://github.com/quantumlib/Qualtran/pull/205). Instead we discuss the superior decomposition propsed by [Berry et al., 2018](https://doi.org/10.1038/s41534-018-0071-5) which has $8n + \\mathcal{O}(1)$ T complexiy and logarithmic depth.\n",
```

Comment:
This seems to be implying that the 8n cost is due to quantum-quantum, when it's actually due to wanting sign(a - b) instead of a < b. A quantum-quantum implementation of a < b is also 4n +- O(1) T cost.

Reply:
fixed